crypto/s2n_certificate.c

```diff
@@ -167,30 +169,131 @@ struct s2n_cert_chain_and_key *s2n_cert_chain_and_key_new(void)
     memset(&chain_and_key->sct_list, 0, sizeof(chain_and_key->sct_list));
-    chain_and_key->san_names = NULL;
-    chain_and_key->x509_cert = NULL;
+    chain_and_key->cn_names = s2n_array_new(sizeof(struct s2n_blob));
+    if (!chain_and_key->cn_names) {
+        return NULL;
+    }
+
+    chain_and_key->san_names = s2n_array_new(sizeof(struct s2n_blob));
+    if (!chain_and_key->san_names) {
+        return NULL;
+    }
 
     return chain_and_key;
 }
 
-static int s2n_cert_chain_and_key_set_x509(struct s2n_cert_chain_and_key *chain_and_key, struct s2n_blob *leaf_bytes)
+static int s2n_cert_chain_and_key_load_sans(struct s2n_cert_chain_and_key *chain_and_key, X509 *x509_cert)
 {
-    const unsigned char *leaf_der = leaf_bytes->data;
-    X509 *cert = d2i_X509(NULL, &leaf_der, leaf_bytes->size);
-    if (!cert) {
-        S2N_ERROR(S2N_ERR_INVALID_PEM);
-    }
+    notnull_check(chain_and_key->san_names);
 
-    chain_and_key->x509_cert = cert;
-
-    GENERAL_NAMES *san_names = X509_get_ext_d2i(cert, NID_subject_alt_name, NULL, NULL);
+    GENERAL_NAMES *san_names = X509_get_ext_d2i(x509_cert, NID_subject_alt_name, NULL, NULL);
     if (san_names == NULL) {
         /* No SAN extension */
         return 0;
     }
 
-    chain_and_key->san_names = san_names;
+    for (int i = 0; i < sk_GENERAL_NAME_num(san_names); i++) {
+        GENERAL_NAME *san_name = sk_GENERAL_NAME_value(san_names, i);
+        if (!san_name) {
+            continue;
+        }
 
+        if (san_name->type == GEN_DNS) {
+            /* Per https://www.openssl.org/docs/man1.1.0/man3/ASN1_STRING_data.html there may
+             * be embedded NULLs inside of the SAN string. The comparison later with any proper
+             * DNS hostnames will safely fail.
+             */
+            unsigned char *san_str = san_name->d.dNSName->data;
+            const size_t san_str_len = san_name->d.dNSName->length;
+            struct s2n_blob *san_blob = s2n_array_insert(chain_and_key->san_names, s2n_array_num_elements(chain_and_key->san_names));
+            if (!san_blob) {
+                GENERAL_NAMES_free(san_names);
+                return -1;
+            }
+
+            if (s2n_alloc(san_blob, san_str_len)) {
+                GENERAL_NAMES_free(san_names);
+                return -1;
+            }
+
+            memcpy_check(san_blob->data, san_str, san_str_len);
+            san_blob->size = san_str_len;
+        }
+    }
+
+    GENERAL_NAMES_free(san_names);
+    return 0;
+}
+
+static int s2n_cert_chain_and_key_load_cn(struct s2n_cert_chain_and_key *chain_and_key, X509 *x509_cert)
+{
+    notnull_check(chain_and_key->cn_names);
+
+    X509_NAME *subject = X509_get_subject_name(x509_cert);
+    if (!subject) {
+        return 0;
+    }
+
+    /* Technically there can be multiple CommonNames entries in the Subject. */
+    int lastpos = -1;
+    while((lastpos = X509_NAME_get_index_by_NID(subject, NID_commonName, lastpos)) >= 0) {
```

Comment:
I'm not a big fan of iterating like this but I couldn't think of a more clear way given the requirements of inputs to `X509_NAME_get_index_by_NID`